api/handlers.go

```diff
@@ -30,9 +30,27 @@ const (
 )
 
 type (
-	// LimitsPublic provides public information of the various limits this
+	// ChallengePublic is the response of GET /login, GET /register,
+	// GET /user/pubkey/register
+	ChallengePublic struct {
+		// Challenge is a hex-encoded representation of the []byte challenge.
+		Challenge string `bson:"challenge" json:"challenge"`
+	}
+	// DownloadsGET is the response of GET /user/downloads
+	DownloadsGET struct {
+		Items    []database.DownloadResponse `json:"items"`
```

Comment:
Why is `database.DownloadResponse` in `database` instead of in here?
We can f/u with this but it seems a bit strange.

Overall I'd avoid `database` types as fields in the request/response types.
Same reason as before. It's just not obvious enough that a change to the persistence type updates the http response.

Reply:
I tried to move them all from `database` to `api` but I've missed this one.

I'm not sure I'm 100% onboard with your general line of thinking. Anyone who sees a `json` annotation on a data type should be immediately aware that this is an API thing. Our storage uses `bson` and our API uses `json` - it shouldn't be a surprise, IMO.

We should, by all means, place any API-only fields in specific types used for API-level communication but I don't see any harm in using the DB-level structs as modules of the API-level structs. Otherwise we'll end up manually deep copying data structures whenever we want to send anything over the API. If this is what's expected then I'll need a much bigger refactoring but I would also like to discuss it live before we go that far.

Reply:
In this particular case, it's probably just because the struct name has the word "Response" in it, making it feel a bit quirky/weird. If it would've been `[]database.Item` it wouldn't have triggered me personally. I think Chris's argument about updating persistence indirectly changing the API response is a really good point. Perhaps we can simply get away with it though by dropping a `NOTE` in the docstring that warns developers any change to the struct has a direct effect on what's returned by the API.

Reply:
By the way, if these structs are your database entities and you return them as-is in the API responses, I actually feel they should live in `database` rather than `api`. So I would actually move all of the other ones back to `database` even. Your database entities belong in the database package right.

Reply:
I only moved types which are only used as responses. Where there were no such types, I created them. So they are not database types. I can't access types in module `api` from module `database` because of circular dependencies.

Reply:
Right the `DownloadResponse` name makes it seem like it's an API response because I'd usually not consider something coming from a DB a response.